File: UVA100/main.py
import os, sys, math, array
global tNums
global cache

def gInit():
    global cache
    cache = [0]*1000000

def treeNplusOne(n):
    if n%2 == 1 :
        return 3*n+1
    else :
        return n>>1

def circleLength(n):
    global cache
    if n == 1:
        return 1
    if n < 1000000 and cache[int(n)] != 0 :
        return cache[int(n)]
    length = 1 + circleLength(treeNplusOne(int(n)))
    if n < 1000000:
        cache[int(n)]=length
    return length


def main():
    global tNums
    gInit()
    clmax=1
    #while True:
    #    line = sys.stdin.readline()
    #    if not line:
    #        break
    for line in sys.stdin:
        tNums = line.split()
        if(len(tNums)==2):
            for x in range(min(int(tNums[0]),int(tNums[1])),max(int(tNums[0]),int(tNums[1])+1)):
                clmax=(max(clmax, circleLength(x)))
            print("%d %d %d" %(int(tNums[0]),int(tNums[1]),clmax))
            clmax=1
                        
if __name__ == "__main__":
    gInit()
    main()
    exit(0)

File: UVA100/main.cpp
#include <sstream>
#include <algorithm>
#include <iostream>
#include <cstring>

#define SIZE 1000000
#define UL unsigned long

using namespace std;

UL cache[SIZE];

int init(){
   memset(cache,0,sizeof(cache));
   return 0;
}

UL treeNplusOne(UL n){
   if((n % 2)==1)
      return 3*n+1;
   else
      return n/2;
}

UL circleLength(UL k){
   int length = 0;
   if(k == 1)
      return 1;
   if((k < SIZE) && cache[int(k)] != 0)
      return cache[k];
   length = 1 + circleLength(treeNplusOne(k));
   if(k < SIZE)
      cache[int(k)]=length;
   return length;
}


int main(){
   /*
   string line;
   while(getline(cin, line)){
      
   }
   */
   init();
   unsigned long a = 0;
   unsigned long b = 0;
   while(cin>>a>>b){
      UL _min = min(a,b);
      UL _max = max(a,b);
      UL clmax=1;
      int idx = 0;
      for(idx = _min;idx<=_max;idx++){
         clmax=(max(clmax,circleLength(idx)));
      }
      cout<<a<<" "<<b<<" "<<clmax<<endl;
   }
   return 0;
}
